File: scala-kata/src/main/scala/org/acmebank/katas/bankaccount/Account.scala
package org.acmebank.katas.bankaccount

import java.time.LocalDateTime
import scala.collection.mutable.ListBuffer
import TransactionType._

object Account {
  def apply(): Account = Account(s"ACCOUNT-${System.currentTimeMillis()}", Money.zero, List.empty)
  def apply(accountId: String): Account = Account(accountId, Money.zero, List.empty)
  def apply(initialBalance: Money): Account = {
    val accountId = s"ACCOUNT-${System.currentTimeMillis()}"
    val transactions = if (initialBalance != Money.zero) {
      List(Transaction(LocalDateTime.now(), initialBalance, DEPOSIT, initialBalance, "Initial deposit"))
    } else {
      List.empty
    }
    Account(accountId, initialBalance, transactions)
  }
  def apply(accountId: String, initialBalance: Money): Account = {
    val transactions = if (initialBalance != Money.zero) {
      List(Transaction(LocalDateTime.now(), initialBalance, DEPOSIT, initialBalance, "Initial deposit"))
    } else {
      List.empty
    }
    Account(accountId, initialBalance, transactions)
  }
}

case class Account(accountId: String, balance: Money, transactions: List[Transaction]) {
  
  def deposit(amount: Money): Account = {
    require(amount != null && !amount.isNegative, "Deposit amount must be positive")
    val newBalance = balance + amount
    val transaction = Transaction(LocalDateTime.now(), amount, DEPOSIT, newBalance)
    copy(balance = newBalance, transactions = transactions :+ transaction)
  }

  def withdraw(amount: Money): Account = {
    require(amount != null && !amount.isNegative, "Withdrawal amount must be positive")
    require(balance >= amount, "Insufficient funds for withdrawal")
    val newBalance = balance - amount
    val transaction = Transaction(LocalDateTime.now(), amount, WITHDRAWAL, newBalance)
    copy(balance = newBalance, transactions = transactions :+ transaction)
  }

  def transferTo(toAccount: Account, amount: Money): (Account, Account) = {
    require(toAccount != null, "Destination account cannot be null")
    require(amount != null && !amount.isNegative, "Transfer amount must be positive")
    require(balance >= amount, "Insufficient funds for transfer")
    
    val fromNewBalance = balance - amount
    val toNewBalance = toAccount.balance + amount
    
    val fromTransaction = Transaction(LocalDateTime.now(), amount, TRANSFER_OUT, fromNewBalance, s"Transfer to ${toAccount.accountId}")
    val toTransaction = Transaction(LocalDateTime.now(), amount, TRANSFER_IN, toNewBalance, s"Transfer from $accountId")
    
    val newFromAccount = copy(balance = fromNewBalance, transactions = transactions :+ fromTransaction)
    val newToAccount = toAccount.copy(balance = toNewBalance, transactions = toAccount.transactions :+ toTransaction)
    
    (newFromAccount, newToAccount)
  }

  def printBalance(): String = 
    s"Account: $accountId | Balance: ${balance.toString} | Date: ${LocalDateTime.now().toString}"

  def printStatement(): String = {
    val statement = new StringBuilder
    statement.append("=== Account Statement ===\n")
    statement.append(s"Account: $accountId\n")
    statement.append(s"Current Balance: ${balance.toString}\n")
    statement.append("=== Transaction History ===\n")
    statement.append("Date | Amount | Balance\n")
    statement.append("-------------------------\n")
    
    transactions.foreach { transaction =>
      statement.append(transaction.toString).append("\n")
    }
    
    statement.toString()
  }

  def printStatement(filteredTransactions: List[Transaction]): String = {
    val statement = new StringBuilder
    statement.append("=== Filtered Account Statement ===\n")
    statement.append(s"Account: $accountId\n")
    statement.append(s"Current Balance: ${balance.toString}\n")
    statement.append("=== Filtered Transaction History ===\n")
    statement.append("Date | Amount | Balance\n")
    statement.append("-------------------------\n")
    
    filteredTransactions.foreach { transaction =>
      statement.append(transaction.toString).append("\n")
    }
    
    statement.toString()
  }

  def getDepositsOnly(): List[Transaction] = StatementFilter.depositsOnly(transactions)
  def getWithdrawalsOnly(): List[Transaction] = StatementFilter.withdrawalsOnly(transactions)
}

File: java-kata/src/main/java/org/acmebank/katas/bankaccount/Account.java
package org.acmebank.katas.bankaccount;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Account {
    private Money balance;
    private final List<Transaction> transactions;
    private final String accountId;

    public Account() {
        this("ACCOUNT-" + System.currentTimeMillis());
    }

    public Account(String accountId) {
        this.accountId = accountId;
        this.balance = Money.zero();
        this.transactions = new ArrayList<>();
    }

    public Account(Money initialBalance) {
        this("ACCOUNT-" + System.currentTimeMillis(), initialBalance);
    }

    public Account(String accountId, Money initialBalance) {
        this.accountId = accountId;
        this.balance = initialBalance;
        this.transactions = new ArrayList<>();
        if (!initialBalance.equals(Money.zero())) {
            addTransaction(LocalDateTime.now(), initialBalance, Transaction.Type.DEPOSIT, "Initial deposit");
        }
    }

    public void deposit(Money amount) {
        if (amount == null || amount.isNegative()) {
            throw new IllegalArgumentException("Deposit amount must be positive");
        }
        this.balance = this.balance.add(amount);
        addTransaction(LocalDateTime.now(), amount, Transaction.Type.DEPOSIT, "");
    }

    public void withdraw(Money amount) {
        if (amount == null || amount.isNegative()) {
            throw new IllegalArgumentException("Withdrawal amount must be positive");
        }
        if (this.balance.isLessThan(amount)) {
            throw new IllegalStateException("Insufficient funds for withdrawal");
        }
        this.balance = this.balance.subtract(amount);
        addTransaction(LocalDateTime.now(), amount, Transaction.Type.WITHDRAWAL, "");
    }

    public void transferTo(Account toAccount, Money amount) {
        if (toAccount == null) {
            throw new IllegalArgumentException("Destination account cannot be null");
        }
        if (amount == null || amount.isNegative()) {
            throw new IllegalArgumentException("Transfer amount must be positive");
        }
        if (this.balance.isLessThan(amount)) {
            throw new IllegalStateException("Insufficient funds for transfer");
        }
        
        // Withdraw from this account
        this.balance = this.balance.subtract(amount);
        addTransaction(LocalDateTime.now(), amount, Transaction.Type.TRANSFER_OUT, 
                      "Transfer to " + toAccount.getAccountId());
        
        // Deposit to destination account
        toAccount.balance = toAccount.balance.add(amount);
        toAccount.addTransaction(LocalDateTime.now(), amount, Transaction.Type.TRANSFER_IN, 
                               "Transfer from " + this.getAccountId());
    }

    public Money getBalance() {
        return this.balance;
    }

    public String getAccountId() {
        return this.accountId;
    }

    public String printBalance() {
        return String.format("Account: %s | Balance: %s | Date: %s", 
                           accountId, balance.toString(), LocalDateTime.now().toString());
    }

    public String printStatement() {
        StringBuilder statement = new StringBuilder();
        statement.append("=== Account Statement ===\n");
        statement.append(String.format("Account: %s\n", accountId));
        statement.append(String.format("Current Balance: %s\n", balance.toString()));
        statement.append("=== Transaction History ===\n");
        statement.append("Date | Amount | Balance\n");
        statement.append("-------------------------\n");
        
        for (Transaction transaction : transactions) {
            statement.append(transaction.toString()).append("\n");
        }
        
        return statement.toString();
    }

    public String printStatement(List<Transaction> filteredTransactions) {
        StringBuilder statement = new StringBuilder();
        statement.append("=== Filtered Account Statement ===\n");
        statement.append(String.format("Account: %s\n", accountId));
        statement.append(String.format("Current Balance: %s\n", balance.toString()));
        statement.append("=== Filtered Transaction History ===\n");
        statement.append("Date | Amount | Balance\n");
        statement.append("-------------------------\n");
        
        for (Transaction transaction : filteredTransactions) {
            statement.append(transaction.toString()).append("\n");
        }
        
        return statement.toString();
    }

    public List<Transaction> getTransactions() {
        return Collections.unmodifiableList(transactions);
    }

    public List<Transaction> getDepositsOnly() {
        return StatementFilter.depositsOnly(transactions);
    }

    public List<Transaction> getWithdrawalsOnly() {
        return StatementFilter.withdrawalsOnly(transactions);
    }

    private void addTransaction(LocalDateTime date, Money amount, Transaction.Type type, String description) {
        Transaction transaction = new Transaction(date, amount, type, this.balance, description);
        transactions.add(transaction);
    }
}
